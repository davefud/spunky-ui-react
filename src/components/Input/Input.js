import React from 'react';
import PropTypes from 'prop-types';
import ProgressBar from '../ProgressBar';
import Icon from '../Icon';
import Label from '../Label';
import { renderToStaticMarkup } from 'react-dom/server';

/** Password input with integrated label, quality tips, and show password toggle. */
class Input extends React.Component {
  constructor(props) {
    super(props);
    
    this.state = {
      showPassword: false
    };

    this.toggleShowPassword = this.toggleShowPassword.bind(this);
  }

  // ES7 class properties
  toggleShowPassword = ev => {

    this.setState(prevState => {
      return { showPassword: !prevState.showPassword };
    });

    if (ev) ev.preventDefault();
  }

  render() {
    const { htmlId, value, label, error, onChange, placeholder, maxLength, showVisibilityToggle, quality, ...props } = this.props;
    const { showPassword } = this.state;
    const svgString = encodeURIComponent(renderToStaticMarkup(<Icon name={showPassword ? 'visibility' : 'visibility_off'} viewBox={24} width={16} height={16} />));
    const dataUri = `url("data:image/svg+xml,${svgString}") center / contain no-repeat`;
    let normalStyle = {
        padding: '10px 35px 10px 10px', 
        width: '235px'
    };
    let errorStyle = {
        padding: '10px 35px 10px 10px', 
        width: '235px',
        border: 'solid 1px red'
    };

    let styles = error ? errorStyle : normalStyle;

    return (
      <div>
        <Label htmlFor={htmlId} label={label} required={true} />
        <div style={{position: 'relative', maxWidth: '235px', cursor: 'pointer' }}>
          <span style={{
            content: "",
            position: 'absolute',
            left: '255px',
            top: 0,
            bottom: 0,
            width: '16px',
            height: '40px',
            background: dataUri
          }} onClick={this.toggleShowPassword} />
          
          <label style={{maxWidth: '250px'}}>
            <input 
              id={htmlId} 
              type={showPassword ? 'text' : 'password'}
              placeholder={placeholder}
              value={value} 
              onChange={onChange} 
              style={styles}
              {...props} />

              {error && <div className='error' style={{color: 'red', fontSize: '.75em', fontWeight: 'bold'}}>{error}</div>}
          </label>
        </div>

        { value.length > 0 && quality && <ProgressBar percent={quality} width={130} height={5} /> }

      </div>
    );
  }
}

Input.propTypes = {
  /** Unique HTML ID. Used for tying label to HTML input. Handy hook for automated testing. */
  htmlId: PropTypes.string.isRequired,

  /** Input name. Recommend setting this to match object's property so a single change handler can be used by convention.*/
  name: PropTypes.string.isRequired,

  /** Password value */
  value: PropTypes.any,

  /** Input label */
  label: PropTypes.string,

  /** Function called when password input value changes */
  onChange: PropTypes.func.isRequired,

  /** Max password length accepted */
  maxLength: PropTypes.number,

  /** Placeholder displayed when no password is entered */
  placeholder: PropTypes.string,

  /** Set to true to show the toggle for displaying the currently entered password */
  showVisibilityToggle: PropTypes.bool,

  /** Display password quality visually via ProgressBar, accepts a number between 0 and 100 */
  quality: PropTypes.number,

  /** Validation error to display */
  error: PropTypes.string
};

Input.defaultProps = {
  maxLength: 50,
  showVisibilityToggle: false,
  label: 'Password'
};

export default Input;
