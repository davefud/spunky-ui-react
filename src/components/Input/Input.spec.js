import React from 'react';
import renderer from 'react-test-renderer';
import Input from './Input';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

test('toggles input type when show/hide password clicked', () => {
    const wrapper = shallow(<Input
        htmlId='test'
        name='test'
        value=''
        onChange={() => {}}
        showVisibilityToggle />
    );

    // Password input should have a type of password initially
    expect(wrapper.find({type: 'password'})).toHaveLength(1);
    expect(wrapper.find({type: 'text'})).toHaveLength(0);

    // simulate a click on the icon
    wrapper.find('span').simulate('click'); // jquery style selectors

    // Password input should have a type of text after clicking eye icon to toggle
    expect(wrapper.find({type: 'password'})).toHaveLength(0);
    expect(wrapper.find({type: 'text'})).toHaveLength(1);
    
});

test('hides password quality by default', () => {
    const tree = renderer.create(<Input
        htmlId='test'
        name='test'
        onChange={() => {}}
        value='IUacac#7wfe' />).toJSON();
    expect(tree).toMatchSnapshot();
});