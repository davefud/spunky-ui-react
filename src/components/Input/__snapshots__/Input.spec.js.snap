// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hides password quality by default 1`] = `
<div>
  <label
    htmlFor="test"
    style={
      Object {
        "display": "block",
      }
    }
  >
    Password
     
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </span>
  </label>
  <div
    style={
      Object {
        "cursor": "pointer",
        "maxWidth": "235px",
        "position": "relative",
      }
    }
  >
    <span
      onClick={[Function]}
      style={
        Object {
          "background": "url(\\"data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2216%22%20height%3D%2216%22%20viewBox%3D%220%200%2024%2024%22%20aria-hidden%3D%22true%22%3E%3Cg%20fill%3D%22currentColor%22%3E%3Cpath%20d%3D%22M12%207c2.76%200%205%202.24%205%205%200%20.65-.13%201.26-.36%201.83l2.92%202.92c1.51-1.26%202.7-2.89%203.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4%200-2.74.25-3.98.7l2.16%202.16C10.74%207.13%2011.35%207%2012%207zM2%204.27l2.28%202.28.46.46C3.08%208.3%201.78%2010.02%201%2012c1.73%204.39%206%207.5%2011%207.5%201.55%200%203.03-.3%204.38-.84l.42.42L19.73%2022%2021%2020.73%203.27%203%202%204.27zM7.53%209.8l1.55%201.55c-.05.21-.08.43-.08.65%200%201.66%201.34%203%203%203%20.22%200%20.44-.03.65-.08l1.55%201.55c-.67.33-1.41.53-2.2.53-2.76%200-5-2.24-5-5%200-.79.2-1.53.53-2.2zm4.31-.78l3.15%203.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z%22%3E%3C%2Fpath%3E%3C%2Fg%3E%3C%2Fsvg%3E\\") center / contain no-repeat",
          "bottom": 0,
          "content": "",
          "left": "255px",
          "position": "absolute",
          "top": 0,
          "width": "16px",
        }
      }
    />
    <label
      style={
        Object {
          "maxWidth": "250px",
        }
      }
    >
      <input
        id="test"
        name="test"
        onChange={[Function]}
        style={
          Object {
            "padding": "10px 35px 10px 10px",
            "width": "235px",
          }
        }
        type="password"
        value="IUacac#7wfe"
      />
    </label>
  </div>
</div>
`;
